components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -0,0 +1,399 @@
+      return;
+    case State::kLoading:
+    case State::kSending:
+      complete_status_ = status;
+      return;
+    case State::kCompleted:
+      destination_url_loader_client_->OnComplete(status);
+      return;
+    case State::kAborted:
+      NOTREACHED();
+      return;
+  }
+  NOTREACHED();
+}
+
+void DeAmpURLLoader::FollowRedirect(
+    const std::vector<std::string>& removed_headers,
+    const net::HttpRequestHeaders& modified_headers,
+    const net::HttpRequestHeaders& modified_cors_exempt_headers,
+    const absl::optional<GURL>& new_url) {
+  // DeAmpURLLoader starts handling the request after
+  // OnReceivedResponse(). A redirect response is not expected.
+  NOTREACHED();
+}
+
+void DeAmpURLLoader::SetPriority(net::RequestPriority priority,
+                                 int32_t intra_priority_value) {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->SetPriority(priority, intra_priority_value);
+}
+
+void DeAmpURLLoader::PauseReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->PauseReadingBodyFromNet();
+}
+
+void DeAmpURLLoader::ResumeReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->ResumeReadingBodyFromNet();
+}
+
+void DeAmpURLLoader::OnBodyReadable(MojoResult) {
+  if (state_ == State::kSending) {
+    // The pipe becoming readable when kSending means all buffered body has
+    // already been sent.
+    ForwardBodyToClient();
+    return;
+  }
+
+  size_t start_size = buffered_body_.size();
+  uint32_t read_bytes = kReadBufferSize;
+  buffered_body_.resize(start_size + read_bytes);
+  auto result = body_consumer_handle_->ReadData(
+      &buffered_body_[0] + start_size, &read_bytes, MOJO_READ_DATA_FLAG_NONE);
+  switch (result) {
+    case MOJO_RESULT_OK:
+      break;
+    case MOJO_RESULT_FAILED_PRECONDITION:
+      // Reading is finished.
+      buffered_body_.resize(start_size);
+      CompleteLoading(std::move(buffered_body_));
+      return;
+    case MOJO_RESULT_SHOULD_WAIT:
+      body_consumer_watcher_.ArmOrNotify();
+      return;
+    default:
+      NOTREACHED();
+      return;
+  }
+  DCHECK_EQ(MOJO_RESULT_OK, result);
+
+  // Check for AMP-ness and find canonical link
+  std::string canonical_link;
+
+  if (de_amp_service->FindCanonicalLinkIfAMP(buffered_body_, &canonical_link)) {
+    const GURL canonical_url(canonical_link);
+    if (!de_amp_service->CheckCanonicalLink(canonical_url)) {
+      VLOG(2) << __func__ << " canonical link check failed " << canonical_url;
+      CompleteLoading(std::move(buffered_body_));
+      return;
+    }
+    VLOG(2) << __func__ << " de-amping and loading " << canonical_url;
+    contents_->GetController().LoadURL(canonical_url, content::Referrer(),
+                                       ui::PAGE_TRANSITION_CLIENT_REDIRECT,
+                                       std::string());
+  } else {
+    // Did not find AMP page and/or canonical link, load original
```

Comment:
Why we drop the referrer here?

Reply:
we're not dropping the referrer right? maybe I'm missing something

Reply:
Sorry, wrong line. I'm asking about passing `content::Referrer(),` to `contents_->GetController().LoadURL`

Reply:
Hm true. I'm not seeing any simple way of getting the original referrer at this point. Any ideas?

Reply:
I could pass down the Request from `BraveContentBrowserClient::CreateURLLoaderThrottles`?

Reply:
as discussed in DM I think we can still call `OnReceiveRedirect` here just like `SignedExchangeLoader`